### PR TITLE
Add GoSec Security Scan

### PR DESCRIPTION
**Pull request by xukaren**
## Motivation
Follow up to issue open-telemetry/oteps#144

[GoSec](https://github.com/securego/gosec) is a static analysis engine which scans go source code for security vulnerabilities. As the project grows and we near GA it might be useful to have a workflow which checks for security vulnerabilities so we can ensure every incremental change is following best development practices. Also passing basic security checks will also make sure that there aren't any glaring issues for our users.

## Changes
This PR adds GoSec security checks to the repo
* After every run the workflow uploads the results to GitHub. Details on the run and security alerts will show up in the security tab of this repo.

**Workflow Triggers**

* daily cron job at 1:30am
* workflow_dispatch (in case maintainers want to trigger a security check manually)

cc @alolita

### CODE REVIEW

File: CHANGELOG.md
Comment (djaglowski) [inline CHANGELOG.md:14]: Can this please be moved to a new section at the top, since it wasn't applied prior to the `0.18.0` release?
```suggestion
## Unreleased

### Added
- GoSec workflow added to GitHub Actions ([PR154](https://github.com/open-telemetry/opentelemetry-log-collection/pull/154))
```

**Review by djaglowski — COMMENTED** [review]
Thank you for adding this.

Just a minor note on the changelog format.

**Comment by tigrannajaryan** [comment]
Shouldn't this check be also perform for all PR builds?

**Comment by xukaren** [comment]
> Shouldn't this check be also perform for all PR builds?

I wasn't sure, since the CodeQL / GoSec Actions most other repos, including [Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/.github/workflows/codeql-analysis.yml#L15), [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/.github/workflows/codeql-analysis.yml#L6), [Javascript](https://github.com/open-telemetry/opentelemetry-js/blob/main/.github/workflows/codeql-analysis.yml#L15), [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/.github/workflows/codeql-analysis.yml#L15) were only using a cron job. The [Collector](https://github.com/open-telemetry/opentelemetry-collector/blob/main/.github/workflows/codeql-analysis.yml#L4) switched over from a cron job to running on PRs. Which trigger would the approvers of this repo prefer?

File: CHANGELOG.md
Comment (xukaren) [reply CHANGELOG.md:17]: Added!

**Comment by jsirianni** [comment]
Can we do on PR **and** recurring against main branch?

Detecting known issues before they are merged is ideal, however, I think it could be possible for issues to be added to the detection engine after merge to main.

**Comment by xukaren** [comment]
Added. How should the GoSec scan errors be addressed?

**Comment by djaglowski** [comment]
@xukaren, I'll make a PR to your branch.

**Comment by djaglowski** [comment]
@xukaren I decided to just extend your PR, and opened #157, which will supersede this one.

File: .github/workflows/gosec.yml
Comment (djaglowski) [inline .github/workflows/gosec.yml:36]: @xukaren This step is failing on #157, where issues have been resolved. It looks like maybe this is a copy/paste from #153. Is there supposed to be an equivalent file generated by `gosec`?

**Comment by djaglowski** [comment]
I'm merging this, including the `-no-fail` flag, which I will remove in #157

**Review by djaglowski — APPROVED** [review]

File: .github/workflows/gosec.yml
Comment (xukaren) [reply .github/workflows/gosec.yml:35]: Yep, sorry for the late reply. Pushed similar changes to my fork and was just testing it with the CI before replying here. Thank you for merging this and keeping it modular with $#157!

**Pull request merged**
